File: pathplannerlib/src/main/java/com/pathplanner/lib/path/ConstraintsZone.java
package com.pathplanner.lib.path;

import org.json.simple.JSONObject;

/**
 * A zone on a path with different kinematic constraints
 *
 * @param minPosition Starting waypoint relative position of the zone
 * @param maxPosition End waypoint relative position of the zone
 * @param constraints The {@link com.pathplanner.lib.path.PathConstraints} to apply within the zone
 */
public record ConstraintsZone(double minPosition, double maxPosition, PathConstraints constraints) {
  /**
   * Create a constraints zone from json
   *
   * @param zoneJson A {@link org.json.simple.JSONObject} representing a constraints zone
   * @return The constraints zone defined by the given json object
   */
  static ConstraintsZone fromJson(JSONObject zoneJson) {
    double minPos = ((Number) zoneJson.get("minWaypointRelativePos")).doubleValue();
    double maxPos = ((Number) zoneJson.get("maxWaypointRelativePos")).doubleValue();
    PathConstraints constraints =
        PathConstraints.fromJson((JSONObject) zoneJson.get("constraints"));
    return new ConstraintsZone(minPos, maxPos, constraints);
  }
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/path/ConstraintsZone.cpp
#include "pathplanner/lib/path/ConstraintsZone.h"
#include <algorithm>

using namespace pathplanner;

ConstraintsZone ConstraintsZone::fromJson(const wpi::util::json &json) {
	double minPos = json.at("minWaypointRelativePos").get_number();
	double maxPos = json.at("maxWaypointRelativePos").get_number();
	PathConstraints constraints = PathConstraints::fromJson(
			json.at("constraints"));

	return ConstraintsZone(minPos, maxPos, constraints);
}
